pkg/sql/distsql_plan_bulk.go

```diff
@@ -95,3 +97,86 @@ func (dsp *DistSQLPlanner) setupAllNodesPlanningTenant(
 	})
 	return planCtx, sqlInstanceIDs, nil
 }
+
+// PhysicalPlanMaker describes a function that makes a physical plan.
+type PhysicalPlanMaker func(context.Context, *DistSQLPlanner) (*PhysicalPlan, *PlanningCtx, error)
+
+// CalculatePlanGrowth returns the number of new or updated sql instances in one
+// physical plan relative to an old one, as well as that number as a fraction of
+// the number in the old one.
+func CalculatePlanGrowth(before, after *PhysicalPlan) (int, float64) {
+	var changed int
+	beforeSpecs := before.GenerateFlowSpecs()
+	afterSpecs := after.GenerateFlowSpecs()
+
+	// How many nodes are in after that are not in before, or are in both but
+	// have changed their spec?
+	for n, afterSpec := range afterSpecs {
+		if beforeSpec, ok := beforeSpecs[n]; !ok {
+			changed++ // not in before at all
+		} else if len(beforeSpec.Processors) != len(afterSpec.Processors) {
+			changed++ // in before with different procs
+		} else {
+			for i := range beforeSpec.Processors {
+				// TODO(dt): add equality check method gen, use it.
+				if afterSpec.Processors[i].Size() != beforeSpec.Processors[i].Size() {
+					changed++ // procs have differnet specs
+					break
+				}
+			}
+		}
+	}
+
+	return changed, float64(changed) / float64(len(beforeSpecs))
+}
+
+// ErrPlanChanged is a sentinel marker error for use to signal a plan changed.
+var ErrPlanChanged = errors.New("physical plan has changed")
+
+// PhysicalPlanChangeChecker returns a function which will periodically call the
+// passed function at the requested interval until the returned channel is
+// closed and compare the plan it returns to the passed initial plan, returning
+// and error if it has changed (as defined by CalculatePlanGrowth) by more than
+// the passed threshold. A threshold of 0 disables it.
+func PhysicalPlanChangeChecker(
+	ctx context.Context,
+	name string,
+	initial *PhysicalPlan,
+	fn PhysicalPlanMaker,
+	execCtx JobExecContext,
```

Comment:
Does it make sense to pass DistSQLPlanner() and not expose full JobExecContext?

Reply:
eh, JobExecContext is in the same package, and this is for use in jobs, so I don't see much benefit in adding a new interface.

Reply:
I disagree.  We should not be passing wider interfaces when narrower will do.
Furthermore, the only reason JobExecContext exists here is because of circular dependencies.

Reply:
Eh, I suppose? Used an interface.